api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirImmunizationServiceImpl.java

```diff
@@ -0,0 +1,106 @@
+import ca.uhn.fhir.rest.param.TokenAndListParam;
+import ca.uhn.fhir.rest.param.TokenParam;
+import org.hl7.fhir.r4.model.Immunization;
+import org.openmrs.Concept;
+import org.openmrs.Obs;
+import org.openmrs.api.EncounterService;
+import org.openmrs.api.ObsService;
+import org.openmrs.module.fhir2.api.FhirImmunizationService;
+import org.openmrs.module.fhir2.api.dao.FhirObservationDao;
+import org.openmrs.module.fhir2.api.search.param.SearchParameterMap;
+import org.openmrs.module.fhir2.api.translators.ImmunizationTranslator;
+import org.openmrs.module.fhir2.api.translators.impl.ImmunizationObsGroupHelper;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+@Component
+public class FhirImmunizationServiceImpl implements FhirImmunizationService {
+	
+	@Autowired
+	private ImmunizationTranslator translator;
+	
+	@Autowired
+	private FhirObservationDao obsDao;
+	
+	@Autowired
+	private ObsService obsService;
+	
+	@Autowired
+	private EncounterService encounterService;
+	
+	@Autowired
+	private ImmunizationObsGroupHelper helper;
+	
+	@Override
+	public Immunization get(String uuid) {
+		return translator.toFhirResource(obsDao.get(uuid));
+	}
+	
+	@Override
+	public Immunization create(Immunization newImmunization) {
+		Obs obs = translator.toOpenmrsType(newImmunization);
+		if (obs.getEncounter().getId() == null) {
+			encounterService.saveEncounter(obs.getEncounter());
+		}
+		obs = obsDao.createOrUpdate(obs);
+		return translator.toFhirResource(obs);
+	}
+	
+	@Override
+	public Immunization update(String uuid, Immunization updatedImmunization) {
+		Obs obs = translator.toOpenmrsType(obsDao.get(uuid), updatedImmunization);
+		obs = obsService.saveObs(obs, "Updated when translating a FHIR Immunization resource.");
+		//		obs = obsDao.createOrUpdate(obs);
+		return translator.toFhirResource(obs);
+	}
+	
+	@Override
+	public Immunization delete(String uuid) {
+		// TODO Simply void the underlying obs
+		throw new UnsupportedOperationException("Deleting a FHIR Immunization resource is currently not supported.");
+	}
+	
+	@Override
+	public Concept getOpenmrsImmunizationConcept() {
+		return helper.concept(immunizationGroupingConcept);
+	}
+	
+	@Override
+	public Collection<Immunization> searchImmunizations(ReferenceAndListParam patientParam, SortSpec sort) {
+		
+		SearchParameterMap searchParams = new SearchParameterMap();
+		searchParams.addParameter(PATIENT_REFERENCE_SEARCH_HANDLER, patientParam);
+		TokenAndListParam conceptParam = new TokenAndListParam();
+		TokenParam token = new TokenParam();
+		token.setValue(getOpenmrsImmunizationConcept().getId().toString());
+		conceptParam.addAnd(token);
+		searchParams.addParameter(CODED_SEARCH_HANDLER, conceptParam);
+		
+		List<String> matchingResourceUuids = obsDao.getSearchResultUuids(searchParams);
+		Collection<Obs> obs = obsDao.getSearchResults(searchParams, matchingResourceUuids);
```

Comment:
I found this in a Spring test, and, I've got an issue with it.
Mostly that `getSearchResults` breaks down with an empty second param.

For now I'm doing some trials to work around that with an ugly if, but surely there's a better way that I don't know of to do something equivalent.

Reply:
You should really be returning a `SearchQueryBundleProvider` from this method to support paging, etc. The `SearchQueryBundleProvider` is also designed to do those DAO calls directly. See [how things our structured in our other service classes](https://github.com/openmrs/openmrs-module-fhir2/blob/4f3ce0524f82735927abf17fc6da615e5b51e1e3/api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirPatientServiceImpl.java#L38).

Reply:
Thanks @ibacher that was a very helpful tip, I did use `SearchQueryBundleProvider` in my next commit 👍